Updates for Beat Saber v1.37.1-2

daniel-chambers opened the pull request:
This PR builds on top of #6 and adds fixes for Beat Saber v1.37.1.

Edit: Also works for v1.37.2.

daniel-chambers committed:
Upgrade project format

daniel-chambers committed:
Use a newer version BSMT.Tasks

This has the targets built in now

daniel-chambers committed:
Reworked LevelStatsViewPatches to be compatible with BS 1.36.0

daniel-chambers committed:
Fixes for BeatSaber v1.37.1

aswild commented:
Thanks @daniel-chambers! It's been a long time since I've played BS but I appreciate the fixes going into keeping my old set of mods working.

One question/concern with compiling this: does it matter that the `version` in `manifest.json` (1.4.17) doesn't match the `AssemblyVersion` and `AssemblyFileVersion` in `AssemblyInfo.cs` (1.4.16)? When building the Release configuration I get a "CompareVersions" error. Bumping the versions in `AssemblyInfo.cs` makes things happy. Looks like in Debug builds this is only a warning.

This is my first time building BS mods so I'm not sure what's normal around here, so to speak.

daniel-chambers committed:
Fix Assembly version so that it matches the manifest.json

daniel-chambers commented:
Oops, yes, you're right those numbers need to match. Fixed.

I'll warn you though, since playing with this a bit, it seems like there's bugs in the way I've updated and fixed the mod and sometimes when selecting maps in-game it just doesn't work and you don't see your percentage underneath the scoreboard. The logs end up with an error like this:

```
[ERROR @ 22:12:51 | UnityEngine] The AssetBundle 'StreamingAssets\BeatmapLevelsData\damage' can't be loaded because another AssetBundle with the same files is already loaded.
[CRITICAL @ 22:12:51 | UnityEngine] NullReferenceException: Object reference not set to an instance of an object
[CRITICAL @ 22:12:51 | UnityEngine] (wrapper dynamic-method) BeatmapDataLoader+<LoadBeatmapDataAsync>d__4.DMD<BeatmapDataLoader+<LoadBeatmapDataAsync>d__4::MoveNext>(BeatmapDataLoader/<LoadBeatmapDataAsync>d__4&)
[CRITICAL @ 22:12:51 | UnityEngine] --- End of stack trace from previous location where exception was thrown ---
[CRITICAL @ 22:12:51 | UnityEngine] ScorePercentage.HarmonyPatches.LevelStatsViewPatches.PostfixShowStatsAsync (ScorePercentage.HarmonyPatches.LevelStatsViewPatches __instance, BeatmapKey beatmapKey, PlayerData playerData) (at D:/Code/BeatSaber/ScorePercentage/HarmonyPatches/LevelStatsViewPatches.cs:69)
[CRITICAL @ 22:12:51 | UnityEngine] System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <51fded79cd284d4d911c5949aff4cb21>:0)
[CRITICAL @ 22:12:51 | UnityEngine] UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at <6ef01846f2094996aa9c77a64b17827d>:0)
[CRITICAL @ 22:12:51 | UnityEngine] UnityEngine.UnitySynchronizationContext.Exec () (at <6ef01846f2094996aa9c77a64b17827d>:0)
[CRITICAL @ 22:12:51 | UnityEngine] UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at <6ef01846f2094996aa9c77a64b17827d>:0)
```

I haven't had time to figure why this happens and what to do about it, I'm afraid. I'm no modding expert myself, I've just poked around inside the BS code and looked at other mods to muddle my way through! 😅

aswild commented:
oop, I hadn't noticed that but probably because I mainly use this mod for the "level completed" screen changes.

Good luck debugging! I know how it feels to just be poking around trying to hack at existing codebases

ibillingsley commented:
I've also noticed that when using Multiplayer+ (from BeatSaberPlus), the stats on the level complete screen are incorrect, as if comparing to a different map.

I think because it's using `_standardLevelDetailViewController.beatmapLevel` which is the last map you selected in the menu, not necessarily the map played in multiplayer.

ibillingsley commented:
I made a few additional changes at https://github.com/ibillingsley/ScorePercentage to fix the multiplayer issue I mentioned. I may have gone a bit overboard with refactoring though.

Also published an experimental build with added dependency on [SongPlayHistory](https://github.com/qe201020335/SongPlayHistory) to take advantage of its caching system, although I'm not sure if it makes much difference.

daniel-chambers commented:
@ibillingsley Ooh, I'll definitely try out your version when I next play Beat Saber (I'm off injured at the moment 😓).

ibillingsley commented:
Hmm actually it still doesn't work in Multiplayer+
edit: Fixed it properly in latest builds

qe201020335 commented:
> Also published an experimental build with added dependency on [SongPlayHistory](https://github.com/qe201020335/SongPlayHistory) to take advantage of its caching system, although I'm not sure if it makes much difference.

I did some quick implementation and ported this mod into SongPlayHistory. SPH already has the score tracking utils in place so it is pretty simple to get all the data needed.
Is this a good idea? I am not sure, but since these two mods are both for showing progression, I thought they would fit together.

Here is the branch https://github.com/qe201020335/SongPlayHistory/tree/score_percentage 
and a build for 1.38.0
[SongPlayHistory-2.1.6-bs1.38.0-21cd175.zip](https://github.com/user-attachments/files/17690708/SongPlayHistory-2.1.6-bs1.38.0-21cd175.zip)

Also made #8 to see if they want to merge or not.

ibillingsley commented:
Merging them sounds good to me.
Yeah it seemed like a waste to parse the beatmap data twice just to get the max score (can be an expensive operation on large maps), which is why I made the alternate version that hooks into SPH instead. But I did it in a pretty hacky way.

qe201020335 commented:
Created a pr for the merge https://github.com/qe201020335/SongPlayHistory/pull/12

I will finish the merge after a refactor on the player data tracking logic

qe201020335 commented:
It is released: https://github.com/qe201020335/SongPlayHistory/releases/tag/2.2.0
